src/Database/LSMTree/Internal/Chunk.hs

```diff
@@ -0,0 +1,98 @@
+module Database.LSMTree.Internal.Chunk
+(
+    -- * Chunks
+    Chunk (Chunk),
+
+    -- * Balers
+    Baler,
+    createBaler,
+    feedBaler,
+    readBalerRemnant
+)
+where
+
+import           Prelude hiding (concat, length, splitAt)
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.STRef.Strict (STRef, newSTRef, readSTRef, writeSTRef)
+import           Data.Vector.Primitive (Vector, concat, length, slice, splitAt)
+import           Data.Word (Word8)
+
+-- * Chunks
+
+-- | A chunk of bytes, typically output during incremental index serialisation.
+newtype Chunk = Chunk (Vector Word8)
+
+-- * Balers
+
+-- Mutable state of a baler.
+data BalerState = BalerState
+                      ![Vector Word8] -- Reverse list of queued blocks
+                      !Int            -- Total size of queued blocks
+
+{-|
+    An object that receives blocks of bytes and repackages them into chunks of
+    preferably a given maximum size.
+-}
+data Baler s = Baler
+                   !Int                  -- Maximum chunk size
+                   !(STRef s BalerState) -- Reference to the mutable state
+
+-- | Creates a new baler.
+createBaler :: Int            -- ^ Maximum chunk size
+            -> ST s (Baler s) -- ^ Creation of the baler
+createBaler maxChunkSize = Baler maxChunkSize <$> newSTRef (BalerState [] 0)
+
+{-|
+    Feeds a baler a block of bytes.
+
+    Bytes received by a bailer are generally queued for later output, but if
+    feeding new bytes makes the queued content exceed the maximum chunk size
+    then as many chunks of maximum size as possible are output.
+-}
+feedBaler :: Vector Word8 -> Baler s -> ST s [Chunk]
+feedBaler block (Baler maxChunkSize stateRef) = do
+    BalerState remnantBlocksRev remnantSize <- readSTRef stateRef
+    let
+
+        totalSize :: Int
+        totalSize = remnantSize + length block
+
+    if totalSize < maxChunkSize
+        then do
+                 writeSTRef stateRef
+                            (BalerState (block : remnantBlocksRev) totalSize)
+                 return []
+        else do
+                 let
+
+                     newRemnantSize :: Int
+                     newRemnantSize = totalSize `mod` maxChunkSize
+
+                     lastOutput, newRemnant :: Vector Word8
+                     (lastOutput, newRemnant)
+                         = splitAt (length block - newRemnantSize) block
+
+                     output :: Vector Word8
+                     output = concat (reverse (lastOutput : remnantBlocksRev))
+
+                     chunks :: [Chunk]
+                     chunks = Chunk                                        <$>
+                              (\ start -> slice start maxChunkSize output) <$>
+                              init [0, maxChunkSize .. totalSize]
```

Comment:
Maybe it would be fine to return a larger chunk if the current block would fill up multiple chunks. I know the compact index would return multiple chunks too, but I wonder if it is necessary

Reply:
That’s a good question. I’d say that it depends on how the output generated during incremental serialization is used. Are there uses that require chunks of a fixed size?

In any case, I think that this question should be further considered but doesn’t really belong to this particular pull request, as it touches more fundamental design issues and affects indexes in general, not just the ordinary index. Would you perhaps like to create an issue that calls for further discussion of this topic and optionally a change of the interfaces and implementations of all indexes?

Reply:
AFAICS none of our code relies on the code being a specific size. https://github.com/IntersectMBO/lsm-tree/issues/303